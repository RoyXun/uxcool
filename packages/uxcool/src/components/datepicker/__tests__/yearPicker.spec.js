import {
  mountPickerFactory, triggerEvent, getPortal, waitTime
} from '@cloud-sn/v-test-utils';
import dayjs from 'dayjs';
import YearPicker from '../yearPicker';
import localeEN from '../locale/en_US';

describe('YearPicker', () => {
  const mountPicker = mountPickerFactory(YearPicker);
  const mockDate = dayjs('2019-05-03');
  const format = 'YYYY';
  it('create', async () => {
    const wrapper = await mountPicker({
      propsData: {
        value: mockDate.toDate(),
        transitionName: '',
      },
    });

    await triggerEvent(wrapper, 'click');

    const portal = await getPortal(wrapper);

    const input = wrapper.find('.ux-input');

    expect(input.element.value).toBe(mockDate.format(format));

    expect(portal.find('.ux-calendar-year-panel-selected-cell').exists()).toBeTruthy();
  });

  describe('value', () => {
    it('init', async () => {
      const initValue = mockDate.subtract(1, 'year');
      const wrapper = await mountPicker({
        propsData: {
          value: initValue.toDate(),
          format,
          transitionName: '',
        },
      });

      const input = wrapper.find('.ux-input');
      expect(input.element.value).toBe(initValue.format(format));
    });
    it('update', async () => {
      const wrapper = await mountPicker({
        propsData: {
          value: mockDate.subtract(1, 'year').toDate(),
          format,
          transitionName: '',
        },
      });

      const updateValue = mockDate.subtract(3, 'year');
      wrapper.setProps({
        value: updateValue.toDate(),
      });
      await waitTime();

      const input = wrapper.find('.ux-input');
      expect(input.element.value).toBe(updateValue.format(format));
    });
  });

  describe('format', () => {
    const customFormat = 'YY';
    it('init', async () => {
      const wrapper = await mountPicker({
        propsData: {
          value: mockDate.toDate(),
          format: customFormat,
          transitionName: '',
        },
      });

      expect(wrapper.find('.ux-input').element.value).toBe(mockDate.format(customFormat));
    });
    it('update', async () => {
      const wrapper = await mountPicker({
        propsData: {
          value: mockDate.toDate(),
          format,
          transitionName: '',
        },
      });

      expect(wrapper.find('.ux-input').element.value).toBe(mockDate.format(format));

      wrapper.setProps({
        format: customFormat,
      });
      await waitTime();

      expect(wrapper.find('.ux-input').element.value).toBe(mockDate.format(customFormat));
    });
  });

  describe('disabledYear', () => {
    function disabledYearFactory(disabledMapping) {
      return (date) => {
        const current = dayjs(date);

        return current.get('year') in disabledMapping;
      };
    }
    it('init and disabled 2013, 2015,2019', async () => {
      const disabledYear = disabledYearFactory({
        2013: 1,
        2015: 1,
        2019: 1,
      });
      const wrapper = await mountPicker({
        propsData: {
          value: mockDate.set('year', 2010).toDate(),
          disabledYear,
          transitionName: '',
        },
      });

      await triggerEvent(wrapper, 'click');

      const portal = await getPortal(wrapper);

      expect(portal.html()).toMatchSnapshot();
    });

    it('update and disabled 2012, 2014, 2018', async () => {
      const disabledYear = disabledYearFactory({
        2012: 1,
        2014: 1,
        2018: 1,
      });
      const wrapper = await mountPicker({
        propsData: {
          value: mockDate.set('year', 2010).toDate(),
          transitionName: '',
        },
      });

      await triggerEvent(wrapper, 'click');
      const portal = await getPortal(wrapper);

      wrapper.setProps({
        disabledYear,
      });

      await waitTime();
      expect(portal.html()).toMatchSnapshot();
    });
  });

  describe('locale', () => {
    it.each([
      ['locale.lang', localeEN.lang],
      ['locale', localeEN],
    ])('should work when use %s', async (_, locale) => {
      const wrapper = await mountPicker({
        propsData: {
          value: mockDate.toDate(),
          locale,
          transitionName: '',
        },
      });

      await triggerEvent(wrapper, 'click');
      const portal = await getPortal(wrapper);

      expect(`${wrapper.html()}${portal.html()}`).toMatchSnapshot();
    });
  });

  describe('decade', () => {
    const mockValue = mockDate.set('year', 2016);
    it('prev decade click', async () => {
      const wrapper = await mountPicker({
        propsData: {
          value: mockValue.toDate(),
          transitionName: '',
        },
      });

      await triggerEvent(wrapper, 'click');

      const portal = await getPortal(wrapper);

      await triggerEvent(portal.find('.ux-calendar-year-panel-prev-decade-btn'), 'click');

      expect(
        portal
          .findAll('.ux-calendar-year-panel-cell')
          .at(1)
          .attributes('title')
      ).toBe('2000');
    });
    it('next decade click', async () => {
      const wrapper = await mountPicker({
        propsData: {
          value: mockValue.toDate(),
          transitionName: '',
        },
      });

      await triggerEvent(wrapper, 'click');

      const portal = await getPortal(wrapper);

      await triggerEvent(portal.find('.ux-calendar-year-panel-next-decade-btn'), 'click');

      expect(
        portal
          .findAll('.ux-calendar-year-panel-cell')
          .at(1)
          .attributes('title')
      ).toBe('2020');
    });

    it('should go to prev panel when select first year', async () => {
      const wrapper = await mountPicker({
        propsData: {
          value: mockValue.toDate(),
          transitionName: '',
        },
      });

      await triggerEvent(wrapper, 'click');
      const portal = await getPortal(wrapper);
      await triggerEvent(portal.find('.ux-calendar-year-panel-last-decade-cell'), 'click');
      expect(
        portal
          .findAll('.ux-calendar-year-panel-cell')
          .at(1)
          .attributes('title')
      ).toBe('2000');
    });
    it('should go to next panel when select last year', async () => {
      const wrapper = await mountPicker({
        propsData: {
          value: mockValue.toDate(),
          transitionName: '',
        },
      });

      await triggerEvent(wrapper, 'click');
      const portal = await getPortal(wrapper);
      await triggerEvent(portal.find('.ux-calendar-year-panel-next-decade-cell'), 'click');
      expect(
        portal
          .findAll('.ux-calendar-year-panel-cell')
          .at(1)
          .attributes('title')
      ).toBe('2020');
    });
  });
});
