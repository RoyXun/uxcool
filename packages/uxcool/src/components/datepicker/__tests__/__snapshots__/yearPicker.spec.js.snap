// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YearPicker disabledYear init and disabled 2013, 2015,2019 1`] = `
<div role="popup">
  <div role="align-popup" class="ux-calendar-picker-container ux-calendar-picker-container-placement-bottomLeft " style="left: -999px; top: -1002px;">
    <div class="ux-calendar-picker-container-content">
      <div class="ux-calendar ux-calendar-month ux-calendar-month-calendar">
        <div class="ux-calendar-month-calendar-content">
          <div class="ux-calendar-month-header-wrap">
            <div class="ux-calendar-header">
              <div style="position: relative;"><a title="上一年 (Control键加左方向键)" role="button" class="ux-calendar-prev-year-btn"></a> <a title="上个月 (翻页上键)" role="button" class="ux-calendar-prev-month-btn"></a> <span class="ux-calendar-ym-select"><a role="button" title="选择年份" class="ux-calendar-year-select">2010年</a><a role="button" title="选择月份" class="ux-calendar-month-select">5月</a></span> <a title="下个月 (翻页下键)" role="button" class="ux-calendar-next-month-btn"></a> <a title="下一年 (Control键加右方向键)" role="button" class="ux-calendar-next-year-btn"></a></div>
              <div>
                <!---->
                <div class="ux-calendar-year-panel">
                  <div class="ux-calendar-year-panel-header"><a title="上一年代" role="button" class="ux-calendar-year-panel-prev-decade-btn"></a> <a class="ux-calendar-year-panel-decade-select"><span class="ux-calendar-year-panel-decade-select-content">
        2010-2019
      </span></a> <a title="下一年代" role="button" class="ux-calendar-year-panel-next-decade-btn"></a></div>
                  <div class="ux-calendar-year-panel-body">
                    <table cellspacing="0" role="grid" class="ux-calendar-year-panel-table">
                      <tbody class="ux-calendar-year-panel-tbody">
                        <tr role="row">
                          <td role="gridcell" title="2009" class="ux-calendar-year-panel-cell ux-calendar-year-panel-last-decade-cell"><a class="ux-calendar-year-panel-year">
                              2009
                            </a></td>
                          <td role="gridcell" title="2010" class="ux-calendar-year-panel-cell ux-calendar-year-panel-selected-cell"><a class="ux-calendar-year-panel-year">
                              2010
                            </a></td>
                          <td role="gridcell" title="2011" class="ux-calendar-year-panel-cell"><a class="ux-calendar-year-panel-year">
                              2011
                            </a></td>
                        </tr>
                        <tr role="row">
                          <td role="gridcell" title="2012" class="ux-calendar-year-panel-cell"><a class="ux-calendar-year-panel-year">
                              2012
                            </a></td>
                          <td role="gridcell" title="2013" class="ux-calendar-year-panel-cell ux-calendar-year-panel-cell-disabled"><a class="ux-calendar-year-panel-year">
                              2013
                            </a></td>
                          <td role="gridcell" title="2014" class="ux-calendar-year-panel-cell"><a class="ux-calendar-year-panel-year">
                              2014
                            </a></td>
                        </tr>
                        <tr role="row">
                          <td role="gridcell" title="2015" class="ux-calendar-year-panel-cell ux-calendar-year-panel-cell-disabled"><a class="ux-calendar-year-panel-year">
                              2015
                            </a></td>
                          <td role="gridcell" title="2016" class="ux-calendar-year-panel-cell"><a class="ux-calendar-year-panel-year">
                              2016
                            </a></td>
                          <td role="gridcell" title="2017" class="ux-calendar-year-panel-cell"><a class="ux-calendar-year-panel-year">
                              2017
                            </a></td>
                        </tr>
                        <tr role="row">
                          <td role="gridcell" title="2018" class="ux-calendar-year-panel-cell"><a class="ux-calendar-year-panel-year">
                              2018
                            </a></td>
                          <td role="gridcell" title="2019" class="ux-calendar-year-panel-cell ux-calendar-year-panel-cell-disabled"><a class="ux-calendar-year-panel-year">
                              2019
                            </a></td>
                          <td role="gridcell" title="2020" class="ux-calendar-year-panel-cell ux-calendar-year-panel-next-decade-cell"><a class="ux-calendar-year-panel-year">
                              2020
                            </a></td>
                        </tr>
                      </tbody>
                    </table>
                  </div>
                </div>
                <!---->
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`YearPicker disabledYear update and disabled 2012, 2014, 2018 1`] = `
<div role="popup">
  <div role="align-popup" class="ux-calendar-picker-container ux-calendar-picker-container-placement-bottomLeft " style="left: -999px; top: -1002px;">
    <div class="ux-calendar-picker-container-content">
      <div class="ux-calendar ux-calendar-month ux-calendar-month-calendar">
        <div class="ux-calendar-month-calendar-content">
          <div class="ux-calendar-month-header-wrap">
            <div class="ux-calendar-header">
              <div style="position: relative;"><a title="上一年 (Control键加左方向键)" role="button" class="ux-calendar-prev-year-btn"></a> <a title="上个月 (翻页上键)" role="button" class="ux-calendar-prev-month-btn"></a> <span class="ux-calendar-ym-select"><a role="button" title="选择年份" class="ux-calendar-year-select">2010年</a><a role="button" title="选择月份" class="ux-calendar-month-select">5月</a></span> <a title="下个月 (翻页下键)" role="button" class="ux-calendar-next-month-btn"></a> <a title="下一年 (Control键加右方向键)" role="button" class="ux-calendar-next-year-btn"></a></div>
              <div>
                <!---->
                <div class="ux-calendar-year-panel">
                  <div class="ux-calendar-year-panel-header"><a title="上一年代" role="button" class="ux-calendar-year-panel-prev-decade-btn"></a> <a class="ux-calendar-year-panel-decade-select"><span class="ux-calendar-year-panel-decade-select-content">
        2010-2019
      </span></a> <a title="下一年代" role="button" class="ux-calendar-year-panel-next-decade-btn"></a></div>
                  <div class="ux-calendar-year-panel-body">
                    <table cellspacing="0" role="grid" class="ux-calendar-year-panel-table">
                      <tbody class="ux-calendar-year-panel-tbody">
                        <tr role="row">
                          <td role="gridcell" title="2009" class="ux-calendar-year-panel-cell ux-calendar-year-panel-last-decade-cell"><a class="ux-calendar-year-panel-year">
                              2009
                            </a></td>
                          <td role="gridcell" title="2010" class="ux-calendar-year-panel-cell ux-calendar-year-panel-selected-cell"><a class="ux-calendar-year-panel-year">
                              2010
                            </a></td>
                          <td role="gridcell" title="2011" class="ux-calendar-year-panel-cell"><a class="ux-calendar-year-panel-year">
                              2011
                            </a></td>
                        </tr>
                        <tr role="row">
                          <td role="gridcell" title="2012" class="ux-calendar-year-panel-cell ux-calendar-year-panel-cell-disabled"><a class="ux-calendar-year-panel-year">
                              2012
                            </a></td>
                          <td role="gridcell" title="2013" class="ux-calendar-year-panel-cell"><a class="ux-calendar-year-panel-year">
                              2013
                            </a></td>
                          <td role="gridcell" title="2014" class="ux-calendar-year-panel-cell ux-calendar-year-panel-cell-disabled"><a class="ux-calendar-year-panel-year">
                              2014
                            </a></td>
                        </tr>
                        <tr role="row">
                          <td role="gridcell" title="2015" class="ux-calendar-year-panel-cell"><a class="ux-calendar-year-panel-year">
                              2015
                            </a></td>
                          <td role="gridcell" title="2016" class="ux-calendar-year-panel-cell"><a class="ux-calendar-year-panel-year">
                              2016
                            </a></td>
                          <td role="gridcell" title="2017" class="ux-calendar-year-panel-cell"><a class="ux-calendar-year-panel-year">
                              2017
                            </a></td>
                        </tr>
                        <tr role="row">
                          <td role="gridcell" title="2018" class="ux-calendar-year-panel-cell ux-calendar-year-panel-cell-disabled"><a class="ux-calendar-year-panel-year">
                              2018
                            </a></td>
                          <td role="gridcell" title="2019" class="ux-calendar-year-panel-cell"><a class="ux-calendar-year-panel-year">
                              2019
                            </a></td>
                          <td role="gridcell" title="2020" class="ux-calendar-year-panel-cell ux-calendar-year-panel-next-decade-cell"><a class="ux-calendar-year-panel-year">
                              2020
                            </a></td>
                        </tr>
                      </tbody>
                    </table>
                  </div>
                </div>
                <!---->
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`YearPicker locale should work when use locale 1`] = `
<div class="ux-calendar-picker ux-calendar-picker-default" style=""><input type="text" readonly="readonly" placeholder="Select year" class="ux-input ux-calendar-picker-input"><i class="fu fu-close_circle ux-calendar-picker-clear"></i><span class="ux-calendar-picker-icon"></span></div>
<div role="popup">
  <div role="align-popup" class="ux-calendar-picker-container ux-calendar-picker-container-placement-bottomLeft " style="left: -999px; top: -1002px;">
    <div class="ux-calendar-picker-container-content">
      <div class="ux-calendar ux-calendar-month ux-calendar-month-calendar">
        <div class="ux-calendar-month-calendar-content">
          <div class="ux-calendar-month-header-wrap">
            <div class="ux-calendar-header">
              <div style="position: relative;"><a title="Last year (Control + left)" role="button" class="ux-calendar-prev-year-btn"></a> <a title="Previous month (PageUp)" role="button" class="ux-calendar-prev-month-btn"></a> <span class="ux-calendar-my-select"><a role="button" title="Choose a month" class="ux-calendar-month-select">May</a><a role="button" title="Choose a year" class="ux-calendar-year-select">2019</a></span> <a title="Next month (PageDown)" role="button" class="ux-calendar-next-month-btn"></a> <a title="Next year (Control + right)" role="button" class="ux-calendar-next-year-btn"></a></div>
              <div>
                <!---->
                <div class="ux-calendar-year-panel">
                  <div class="ux-calendar-year-panel-header"><a title="Last decade" role="button" class="ux-calendar-year-panel-prev-decade-btn"></a> <a class="ux-calendar-year-panel-decade-select"><span class="ux-calendar-year-panel-decade-select-content">
        2010-2019
      </span></a> <a title="Next decade" role="button" class="ux-calendar-year-panel-next-decade-btn"></a></div>
                  <div class="ux-calendar-year-panel-body">
                    <table cellspacing="0" role="grid" class="ux-calendar-year-panel-table">
                      <tbody class="ux-calendar-year-panel-tbody">
                        <tr role="row">
                          <td role="gridcell" title="2009" class="ux-calendar-year-panel-cell ux-calendar-year-panel-last-decade-cell"><a class="ux-calendar-year-panel-year">
                              2009
                            </a></td>
                          <td role="gridcell" title="2010" class="ux-calendar-year-panel-cell"><a class="ux-calendar-year-panel-year">
                              2010
                            </a></td>
                          <td role="gridcell" title="2011" class="ux-calendar-year-panel-cell"><a class="ux-calendar-year-panel-year">
                              2011
                            </a></td>
                        </tr>
                        <tr role="row">
                          <td role="gridcell" title="2012" class="ux-calendar-year-panel-cell"><a class="ux-calendar-year-panel-year">
                              2012
                            </a></td>
                          <td role="gridcell" title="2013" class="ux-calendar-year-panel-cell"><a class="ux-calendar-year-panel-year">
                              2013
                            </a></td>
                          <td role="gridcell" title="2014" class="ux-calendar-year-panel-cell"><a class="ux-calendar-year-panel-year">
                              2014
                            </a></td>
                        </tr>
                        <tr role="row">
                          <td role="gridcell" title="2015" class="ux-calendar-year-panel-cell"><a class="ux-calendar-year-panel-year">
                              2015
                            </a></td>
                          <td role="gridcell" title="2016" class="ux-calendar-year-panel-cell"><a class="ux-calendar-year-panel-year">
                              2016
                            </a></td>
                          <td role="gridcell" title="2017" class="ux-calendar-year-panel-cell"><a class="ux-calendar-year-panel-year">
                              2017
                            </a></td>
                        </tr>
                        <tr role="row">
                          <td role="gridcell" title="2018" class="ux-calendar-year-panel-cell"><a class="ux-calendar-year-panel-year">
                              2018
                            </a></td>
                          <td role="gridcell" title="2019" class="ux-calendar-year-panel-cell ux-calendar-year-panel-selected-cell"><a class="ux-calendar-year-panel-year">
                              2019
                            </a></td>
                          <td role="gridcell" title="2020" class="ux-calendar-year-panel-cell ux-calendar-year-panel-next-decade-cell"><a class="ux-calendar-year-panel-year">
                              2020
                            </a></td>
                        </tr>
                      </tbody>
                    </table>
                  </div>
                </div>
                <!---->
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`YearPicker locale should work when use locale.lang 1`] = `
<div class="ux-calendar-picker ux-calendar-picker-default" style=""><input type="text" readonly="readonly" placeholder="Select year" class="ux-input ux-calendar-picker-input"><i class="fu fu-close_circle ux-calendar-picker-clear"></i><span class="ux-calendar-picker-icon"></span></div>
<div role="popup">
  <div role="align-popup" class="ux-calendar-picker-container ux-calendar-picker-container-placement-bottomLeft " style="left: -999px; top: -1002px;">
    <div class="ux-calendar-picker-container-content">
      <div class="ux-calendar ux-calendar-month ux-calendar-month-calendar">
        <div class="ux-calendar-month-calendar-content">
          <div class="ux-calendar-month-header-wrap">
            <div class="ux-calendar-header">
              <div style="position: relative;"><a title="Last year (Control + left)" role="button" class="ux-calendar-prev-year-btn"></a> <a title="Previous month (PageUp)" role="button" class="ux-calendar-prev-month-btn"></a> <span class="ux-calendar-my-select"><a role="button" title="Choose a month" class="ux-calendar-month-select">May</a><a role="button" title="Choose a year" class="ux-calendar-year-select">2019</a></span> <a title="Next month (PageDown)" role="button" class="ux-calendar-next-month-btn"></a> <a title="Next year (Control + right)" role="button" class="ux-calendar-next-year-btn"></a></div>
              <div>
                <!---->
                <div class="ux-calendar-year-panel">
                  <div class="ux-calendar-year-panel-header"><a title="Last decade" role="button" class="ux-calendar-year-panel-prev-decade-btn"></a> <a class="ux-calendar-year-panel-decade-select"><span class="ux-calendar-year-panel-decade-select-content">
        2010-2019
      </span></a> <a title="Next decade" role="button" class="ux-calendar-year-panel-next-decade-btn"></a></div>
                  <div class="ux-calendar-year-panel-body">
                    <table cellspacing="0" role="grid" class="ux-calendar-year-panel-table">
                      <tbody class="ux-calendar-year-panel-tbody">
                        <tr role="row">
                          <td role="gridcell" title="2009" class="ux-calendar-year-panel-cell ux-calendar-year-panel-last-decade-cell"><a class="ux-calendar-year-panel-year">
                              2009
                            </a></td>
                          <td role="gridcell" title="2010" class="ux-calendar-year-panel-cell"><a class="ux-calendar-year-panel-year">
                              2010
                            </a></td>
                          <td role="gridcell" title="2011" class="ux-calendar-year-panel-cell"><a class="ux-calendar-year-panel-year">
                              2011
                            </a></td>
                        </tr>
                        <tr role="row">
                          <td role="gridcell" title="2012" class="ux-calendar-year-panel-cell"><a class="ux-calendar-year-panel-year">
                              2012
                            </a></td>
                          <td role="gridcell" title="2013" class="ux-calendar-year-panel-cell"><a class="ux-calendar-year-panel-year">
                              2013
                            </a></td>
                          <td role="gridcell" title="2014" class="ux-calendar-year-panel-cell"><a class="ux-calendar-year-panel-year">
                              2014
                            </a></td>
                        </tr>
                        <tr role="row">
                          <td role="gridcell" title="2015" class="ux-calendar-year-panel-cell"><a class="ux-calendar-year-panel-year">
                              2015
                            </a></td>
                          <td role="gridcell" title="2016" class="ux-calendar-year-panel-cell"><a class="ux-calendar-year-panel-year">
                              2016
                            </a></td>
                          <td role="gridcell" title="2017" class="ux-calendar-year-panel-cell"><a class="ux-calendar-year-panel-year">
                              2017
                            </a></td>
                        </tr>
                        <tr role="row">
                          <td role="gridcell" title="2018" class="ux-calendar-year-panel-cell"><a class="ux-calendar-year-panel-year">
                              2018
                            </a></td>
                          <td role="gridcell" title="2019" class="ux-calendar-year-panel-cell ux-calendar-year-panel-selected-cell"><a class="ux-calendar-year-panel-year">
                              2019
                            </a></td>
                          <td role="gridcell" title="2020" class="ux-calendar-year-panel-cell ux-calendar-year-panel-next-decade-cell"><a class="ux-calendar-year-panel-year">
                              2020
                            </a></td>
                        </tr>
                      </tbody>
                    </table>
                  </div>
                </div>
                <!---->
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
